common/fitparser.cpp

```diff
@@ -0,0 +1,1090 @@
+                        (void)parseFitEntryBootGuardBootPolicy(item, localOffset, itemIndex, info, currentEntrySize);
+                        bpIndex = itemIndex;
+                        break;
+                        
+                    default:
+                        // Do nothing
+                        break;
+                }
+            }
+            else {
+                msg(usprintf("%s: FIT entry #%u not found in the image", __FUNCTION__, i), fitIndex);
+            }
+        }
+        
+        // Explicitly set the item referenced by FIT as fixed
+        if (itemIndex.isValid()) {
+            model->setFixed(itemIndex, true);
+        }
+        
+        // Add entry to fitTable
+        currentStrings.push_back(usprintf("%016" PRIX64 "h", currentEntry->Address));
+        currentStrings.push_back(usprintf("%08Xh", currentEntrySize));
+        currentStrings.push_back(usprintf("%04Xh", currentEntry->Version));
+        currentStrings.push_back(usprintf("%02Xh", currentEntry->Checksum));
+        currentStrings.push_back(fitEntryTypeToUString(currentEntry->Type));
+        currentStrings.push_back(info);
+        fitTable.push_back(std::pair<std::vector<UString>, UModelIndex>(currentStrings, itemIndex));
+    }
+    
+    // Perform validation of BootGuard components
+    if (bgAcmFound) {
+        if (!bgKeyManifestFound) {
+            msg(usprintf("%s: startup ACM found, but KeyManifest is not", __FUNCTION__), acmIndex);
+        }
+        else if (!bgBootPolicyFound) {
+            msg(usprintf("%s: startup ACM and Key Manifest found, Boot Policy is not", __FUNCTION__), kmIndex);
+        }
+        else {
+            // Check key hashes
+            if (!bgKmHash.isEmpty()
+                && !(bgKmHash == bgBpHashSha256 || bgKmHash == bgBpHashSha384)) {
+                msg(usprintf("%s: Boot Policy key hash stored in Key Manifest differs from the hash of the public key stored in Boot Policy", __FUNCTION__), bpIndex);
+                return U_SUCCESS;
+            }
+        }
+    }
+
+    return U_SUCCESS;
+}
+
+void FitParser::findFitRecursive(const UModelIndex & index, UModelIndex & found, UINT32 & fitOffset)
+{
+    // Sanity check
+    if (!index.isValid()) {
+        return;
+    }
+    
+    // Process child items
+    for (int i = 0; i < model->rowCount(index); i++) {
+        findFitRecursive(index.model()->index(i, 0, index), found, fitOffset);
+        
+        if (found.isValid()) {
+            // Found it, no need to process further
+            return;
+        }
+    }
+    
+    // Check for all FIT signatures in item body
+    UByteArray lastVtfBody = model->body(ffsParser->lastVtf);
+    UINT64 fitSignatureValue = INTEL_FIT_SIGNATURE;
+    UByteArray fitSignature((const char*)&fitSignatureValue, sizeof(fitSignatureValue));
+    UINT32 storedFitAddress = *(const UINT32*)(lastVtfBody.constData() + lastVtfBody.size() - INTEL_FIT_POINTER_OFFSET);
```

Comment:
Is this guaranteed to produce valid address?

Reply:
Yes, as the minimal size of UEFI image (volume header + ffs file header) is more than 0x40 (the value of INTEL_FIT_POINTER_OFFSET).